### PR TITLE
Add B3 propagator

### LINKED ISSUE #853: Add b3 propagator
Add support for B3 propagator that converts SpanContext to/from x-b3-* Headers, based on https://github.com/openzipkin/b3-propagation

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
Fixes #853.

Adds support for B3 propagation, including both single and multi header formats.

### CODE REVIEW

**Comment by MikeGoldsmith** [comment]
How do I request a new coverage report?

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  public B3Propagator() {
+    this(false);
+  }
+
+  /**
+   * Creates a new instance of {@link B3Propagator}.
+   *
+   * @param singleHeader whether to use single or multiple headers.
+   */
+  public B3Propagator(Boolean singleHeader) {
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: does this need to be a nullable big-b `Boolean`? Can't it just be a primitive boolean?

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  @Override
+  @SuppressWarnings("StringSplitter")
+  public <C /*>>> extends @NonNull Object*/> SpanContext extract(C carrier, Getter<C> getter) {
+    checkNotNull(carrier, "carrier");
+    checkNotNull(getter, "getter");
+
+    String traceId;
+    String spanId;
+    String sampled;
+    if (singleHeader) {
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: This should be broken up into at least 2 separate methods. This method is definitely too long. One method for the single header case that returns the context, and one for the multi-header case that returns the context will make this code much easier to read. Thanks!

Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: I support this. The TraceContext one is refactored as it was too long and it was (relatively) hard to read.

**Comment by carlosalberto** [comment]
Hey @MikeGoldsmith thanks for the PR, it looks very neat and clean. Left a small comment supporting @jkwatson's suggestion to split the extraction method.

One thing I remember is that B3 is expected to handle case insensitive headers for extraction, i.e. use either. `X-B3-TraceId` or. `x-b3-traceid`. At least that's what a few vendors support (DDog and LS for sure). 

The second one is about the single/multiple mode. I'd imagine some users wanting to have both modes supported (by trying the single header case first, falling back to the multiple headers case). This would be specially useful in case they interact with other clients that might send single or multiple modes. wWe could do that in a follow up PR if/as needed though ;)

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+          traceFlags,
+          TraceState.getDefault());
+    } catch (Exception e) {
+      logger.log(Level.INFO, "Error parsing B3 header. Returning INVALID span context.", e);
+      return SpanContext.getInvalid();
+    }
+  }
+
+  private static boolean isTraceIdValid(String value) {
+    return !(value == null || value.isEmpty() || value.length() > MAX_TRACE_ID_LENGTH);
```
Comment (thisthat) [inline api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: The first two checks could leverage the `StringUtils.isNullOrEmpty(value)` method

Comment (MikeGoldsmith) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: Thanks - I've pushed an update to use `StringUtils`.

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java (hunk 1 above)
Comment (MikeGoldsmith) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: Pushed an update - there is no need for a `Boolean` in this case.

**Comment by MikeGoldsmith** [comment]
@carlosalberto I've refactored the extract into two separate methods and addressed the other feedback.

Other implementations of B3 are case-insensitive, but I don't believe any of the OTel version are. This is because we are unable to `get` a key regardless of casing, or retrieve all known keys and compare locally. In OpenTracing with ITextMap, it was possible to get all keys and do a local case insensitive lookup.

I've created issue #985 to track making the B3 propagator automatically switch between single and multi headers.

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  }
+
+  private static SpanContext buildSpanContext(String traceId, String spanId, String sampled) {
+    try {
+      TraceFlags traceFlags =
+          TraceFlags.builder()
+              .setIsSampled(
+                  TRUE_INT.equals(sampled)
+                      || Boolean.parseBoolean(sampled)) // accept either "1" or "true"
+              .build();
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: You can construct a sampled/notSampled TraceFlags and reuse them.

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+      return SpanContext.createFromRemoteParent(
+          TraceId.fromLowerBase16(traceId, 0),
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: This requires traceId (string) to be 32 hex characters and we only check that the size is not greater than MAX_TRACE_ID_LENGTH.

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Implementation of the B3 propagation protocol. See <a
+ * href=https://github.com/openzipkin/b3-propagation>openzipkin/b3-propagation</a>.
+ */
+@Immutable
+public class B3Propagator implements HttpTextFormat<SpanContext> {
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: Should we have 2 propagators: legacy + single-header? I think that simplifies the code. We can do it later.

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    String spanId = parts[1];
+    if (!isSpanIdValid(spanId)) {
+      logger.info(
+          "Invalid SpanId in B3 header: " + COMBINED_HEADER + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String sampled = parts.length == 3 ? parts[2] : null;
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: Based on my understanding sampled is present if parts.length is 3 or 4 :)

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+              + sampled);
+    } else {
+      setter.set(carrier, TRACE_ID_HEADER, spanContext.getTraceId().toLowerBase16());
+      setter.set(carrier, SPAN_ID_HEADER, spanContext.getSpanId().toLowerBase16());
+      setter.set(carrier, SAMPLED_HEADER, sampled);
+    }
+  }
+
+  @Override
+  public <C /*>>> extends @NonNull Object*/> SpanContext extract(C carrier, Getter<C> getter) {
```
[the remark this answers is not in the document]
Comment (MikeGoldsmith) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: Removed :+1:

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java (hunk 6 above)
Comment (MikeGoldsmith) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: Created #999 to track breaking single and multiple header implementations.

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java (hunk 7 above)
Comment (MikeGoldsmith) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: You're correct. We need to check the sampled string if we have three or more items in the array.

File: api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java (hunk 5 above)
Comment (MikeGoldsmith) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: Pushed change to left pad TraceIds upto max with leading 0's. I added padLeft to `StringUtils`. Let me know what you think.

Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java]: As this was fixed along with a test (see https://github.com/open-telemetry/opentelemetry-java/pull/979/files#diff-004fb9fadd7caedae3844dfe8b040186R137) I'm marking this as resolved.

**Pull request merged**